### PR TITLE
Setup publishing

### PR DESCRIPTION
**Pull request by Baccata**
The goal of this is to have access to a published version of the LSP to maven central to facilitate consumption in IDEs.  

Publishing occurs upon merges to the dss branch, or upon tagged release. 

* Adds a mill build (used only for the sake of publishing) 
* Adds dynver-like versioning logic 
* Adds a publish script 
* Adds necessary CI setup 

Publishing is setup correctly, as proven by [this build step](https://github.com/disneystreaming/smithy-language-server/runs/4273043861?check_suite_focus=true). 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -26,3 +26,42 @@ jobs:
...
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.mill
+            ~/.ivy2/cache
+            ~/.coursier/cache/v1
+            ~/.cache/coursier/v1
+            ~/AppData/Local/Coursier/Cache/v1
+            ~/Library/Caches/Coursier/v1
+          key: mill-cache-v2-${{ hashFiles('**/*.sc') }}-${{ hashFiles('.mill-version') }}
```
Comment (keynmol) [inline .github/workflows/ci.yml]: I thought coursier cache supports those folders - or is it just for the mill version? https://github.com/coursier/cache-action#mill

File: build.sc [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    if (gitDirty) sys.error("Dirty workspace !")
+
+    val commitHash =
+      os.proc("git", "rev-parse", "--short", "HEAD").call().out.lines.head.trim
+
+    val describeResult = os
+      .proc(
+        "git",
+        "submodule",
```
Comment (keynmol) [inline build.sc]: Is this right?

File: build.sc [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    val branch = mainBranch()
+    os.proc("git", "describe", branch, "--abbrev=0", "--tags")
+      .call()
+      .out
+      .lines
+      .headOption
+  }
+
+  object int {
+    def unapply(s: String) = Try(s.toInt).toOption
```
Comment (keynmol) [inline build.sc]: Nit: I thought Mill is already 2.13 so we can use `toIntOption`

File: build.sc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    organization = "com.disneystreaming.smithy",
+    url = "https://github.com/disneystreaming/smithy-language-server",
+    licenses = Seq(License.`Apache-2.0`),
+    versionControl = VersionControl(
+      Some("https://github.com/disneystreaming/smithy-language-server")
+    ),
+    Seq(Developer("baccata", "Olivier Mélois", "https://github.com/baccata"))
+  )
+
+  def gitVersion: T[String] = T.input {
```
Comment (keynmol) [inline build.sc:31]: Can https://github.com/lefou/mill-vcs-version be used instead?

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Baccata) [reply .github/workflows/ci.yml]: Not sure, I've blindly copied it from elsewhere 😅

File: build.sc (hunk 4 above)
Comment (Baccata) [reply build.sc:31]: I'll happily switch as soon as Tobias adds MIMA on mill (which in theory is the next big item for him)

File: build.sc (hunk 3 above)
Comment (Baccata) [reply build.sc]: src/main/java/software/amazon/smithy/lsp/SmithyProject.java

Comment (Baccata) [reply build.sc]: 8a04add

File: build.sc (hunk 2 above)
Comment (Baccata) [reply build.sc]: 8a04add

File: .github/workflows/ci.yml (hunk 1 above)
Comment (Baccata) [reply .github/workflows/ci.yml]: Nice : 36d051a

**Review by keynmol — APPROVED** [review]
LGTM

**Pull request merged**